packages/langium/src/documents/document-builder.ts

```diff
@@ -126,6 +161,25 @@ export class DefaultDocumentBuilder implements DocumentBuilder {
             await interruptAndCheck(cancelToken);
             await callback(document);
         }
+        await this.notifyBuildPhase(documents, targetState, cancelToken);
+    }
+
+    onBuildPhase(targetState: DocumentState, callback: DocumentBuildListener): void {
+        if (this.buildPhaseListeners.has(targetState)) {
+            this.buildPhaseListeners.get(targetState)!.push(callback);
+        } else {
+            this.buildPhaseListeners.set(targetState, [callback]);
+        }
```

Comment:
We do this quite often throughout our codebase. Would it make sense to add a `MultiMap` to our utilities? Doesn't have to happen in this PR, but we should create an issue for this.

Reply:
Maybe one 'utility' function would be sufficient?

```ts
// collections.ts
export function insert<K, V>(map: Map<K, V[]>, key: K, value: V): Map<K, V[]> {
    map.has(key) ? map.get(key).push(value) : map.set(key, [value]);
    return map;
}
```

Reply:
Well, that leaves the issue of the `get` function to return `undefined` instead of an empty array, which would be preferable.

Reply:
Makes sense - a `MultiMap` type would be nice.